enhancements/network/ingress-fault-detection.md

```diff
@@ -0,0 +1,163 @@
+
+To satifsy this story, an Ingress debug container that diagnoses and triages Ingress and other network issues will be avaiable to cluster admins and support engineers.
+
+### Implementation Details/Notes/Constraints
+
+Implementing this enhancement requires changes in the following repositories:
+
+* `openshift/cluster-ingress-operator`
+* `openshift/oc`
+
+Proper Prometheus Alert rules will need to be created once the Ingress Operator exposes fault check metrics.
+These rules could live in the Ingress Operator's repository, but will most likely need to be reviewed by a member of the OpenShift monitoring or OpenShift service delivery team.
+
+As is, the Ingress Operator automatic fault detection code is always on. Cluster administrators will not be able to turn off the background check. The thinking here was that
+the "knob" to disable this feature could be provided after this enhancement is released should it be deemed necessary based on customer input.
+
+### Risks and Mitigations
+
+The Ingress debug container should only be accesible to cluster administrators or OpenShift support engineers, to prevent unintended privellege escalations for normal cluster users.
+The Ingress debug container should also not leak any report data to unprivelleged cluster users or unknown users external to the cluster.
+
+## Design Details
+
+### Open Questions
+
+* How should the Ingress debug tool be invoked by cluster administrators? Current thinking is through some sub-command of `oc`.
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Graduation Criteria
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, the Ingress Operator will create the additional resources needed to perform the
+periodic Ingress fault detection. The Ingress Operator will have the ability to update these resources
+should they change between minor release versions.
+
+On downgrade, the background Ingress fault detection echo container will be deleted. Fault detection route and service
+resources may still exist in the `openshift-ingress-operator` namespace, but this is not detrimental, and they can be deleted
+if they are causing issues with regards to HAProxy metrics, etc.
+
+The Ingress debug container is launched on-demand, so there are no upgrade concerns with that component.
+
+### Version Skew Strategy
+
+N/A
```

Comment:
Will the debug container depend on particular cluster-version-linked details?  E.g. what happens if you launch a 4.7 debug container into a 4.1 cluster?  Or a 4.20 cluster where some aspects of the ingress/router system have changed beyond recognition?  Or if a cloud provider has grown a new LB flavor and the 4.20 clusters have pivoted to using that new flavor to back routes?  I expect in most of these situations, the behavior we want is "debug tool gracefully errors out and says 'you might want to use a debug tool extracted from the cluster you're inspecting'" or some such.  Are we actually going to distribute the debug tool as an image referenced from the release image, like we distribute the must-gather, tools, etc. images today?

Reply:
If the debug container were to be distributed as a part of the release image, that would solve the issue of version skew. The `oc` cli debug container invocation could use the debug container included in the current release image, and if none is available in the release image, gracefully exit. Thoughts?